shared/utility/math/filter/eigen/UKF.h

```diff
@@ -0,0 +1,352 @@
+                for (unsigned int i = 0; i < NUM_SIGMA_POINTS; ++i) {
+                    predictions(i) = model.predict(sigma_points.col(i), params...);
+                }
+
+                // Now calculate the mean of these measurement sigmas.
+                MeasurementScalar predicted_mean = mean_from_sigmas(predictions, mean_weights).x();
+                Eigen::Matrix<MeasurementScalar, 1, NUM_SIGMA_POINTS> centred =
+                    (predictions.array() - predicted_mean).matrix();
+
+                // Create a state update in our measurement units
+                Eigen::Matrix<MeasurementScalar, NUM_SIGMA_POINTS, NUM_SIGMA_POINTS> update =
+                    covariance_update.template cast<MeasurementScalar>();
+                update = update.transpose() * centred.transpose()
+                         * (MeasurementScalar(1) / (measurement_variance + centred * update * centred.transpose()))
+                         * centred * update;
+                covariance_update -= update.template cast<Scalar>();
+
+                MeasurementScalar innovation = model.difference(measurement, predicted_mean);
+                d += (centred.transpose() * (MeasurementScalar(1) / measurement_variance) * innovation)
+                         .template cast<Scalar>();
+
+                // Update our mean and covariance
+                mean       = sigma_mean + centred_sigma_points * covariance_update * d;
+                mean       = model.limit(mean);
+                covariance = centred_sigma_points * covariance_update * centred_sigma_points.transpose();
+
+                // Calculate and return the likelihood of the prior mean and covariance given the new measurement
+                // (i.e. the prior probability density of the measurement):
+                return utility::support::LazyEvaluation<MeasurementScalar>(
+                    [predictions, predicted_mean, measurement_variance, innovation, cw = this->covariance_weights] {
+                        MeasurementScalar predicted_covariance =
+                            covariance_from_sigmas(predictions, predicted_mean, cw).x();
+
+                        MeasurementScalar innovation_variance = predicted_covariance + measurement_variance;
+
+                        MeasurementScalar likelihood_exponent =
+                            ((innovation * (MeasurementScalar(1) / innovation_variance)) * innovation);
+
+                        MeasurementScalar loglikelihood =
+                            0.5 * (innovation_variance + std::abs(likelihood_exponent) + std::log(2 * M_PI));
+
+                        return -loglikelihood;
+                    });
+            }
+
+            const StateVec& get() const {
+                return mean;
+            }
+
+            const StateMat& getCovariance() const {
+                return covariance;
+            }
+
+            EIGEN_MAKE_ALIGNED_OPERATOR_NEW
```

Comment:
Not needed in c++17 but I suppose it doesn't hurt to leave it

Reply:
With the patch in the other branch I am pretty sure this is an empty macro now, but we should remove it once everything is merged in